libsatadrivers/include/satadrivers/ide.h

```diff
@@ -0,0 +1,148 @@
+ *
+ */
+
+#pragma once
+
+#include <platsupport/io.h>
+#include <satadrivers/common.h>
+
+// Channels:
+#define ATA_PRIMARY     0x00
+#define ATA_SECONDARY   0x01
+
+// Directions:
+#define ATA_READ        0x00
+#define ATA_WRITE       0x01
+#define ATA_IDENTIFY    0x02
+
+#define ATA_MASTER      0x00
+#define ATA_SLAVE       0x01
+
+#define DMA_DISABLE     0x00
+#define DMA_ENABLE      0x01
+
+// IDE Status
+#define ATA_SR_BSY      0x80    // Busy
+#define ATA_SR_DRDY     0x40    // Drive ready
+#define ATA_SR_DF       0x20    // Drive write fault
+#define ATA_SR_DSC      0x10    // Drive seek complete
+#define ATA_SR_DRQ      0x08    // Data request ready
+#define ATA_SR_CORR     0x04    // Corrected data
+#define ATA_SR_IDX      0x02    // Inlex
+#define ATA_SR_ERR      0x01    // Error
+
+// IDE Errors
+#define ATA_ER_BBK      0x80    // Bad sector
+#define ATA_ER_UNC      0x40    // Uncorrectable data
+#define ATA_ER_MC       0x20    // No media
+#define ATA_ER_IDNF     0x10    // ID mark not found
+#define ATA_ER_MCR      0x08    // No media
+#define ATA_ER_ABRT     0x04    // Command aborted
+#define ATA_ER_TK0NF    0x02    // Track 0 not found
+#define ATA_ER_AMNF     0x01    // No address mark
+
+// IDE Identification offsets
+#define ATA_IDENT_DEVICETYPE   0
+#define ATA_IDENT_CYLINDERS    2
+#define ATA_IDENT_HEADS        6
+#define ATA_IDENT_SECTORS      12
+#define ATA_IDENT_SERIAL       20
+#define ATA_IDENT_MODEL        54
+#define ATA_IDENT_CAPABILITIES 98
+#define ATA_IDENT_FIELDVALID   106
+#define ATA_IDENT_MAX_LBA      120
+#define ATA_IDENT_COMMANDSETS  164
+#define ATA_IDENT_MAX_LBA_EXT  200
+
+// IDE Commands
+#define ATA_CMD_READ_PIO          0x20
+#define ATA_CMD_READ_PIO_EXT      0x24
+#define ATA_CMD_READ_DMA          0xC8
+#define ATA_CMD_READ_DMA_EXT      0x25
+#define ATA_CMD_WRITE_PIO         0x30
+#define ATA_CMD_WRITE_PIO_EXT     0x34
+#define ATA_CMD_WRITE_DMA         0xCA
+#define ATA_CMD_WRITE_DMA_EXT     0x35
+#define ATA_CMD_CACHE_FLUSH       0xE7
+#define ATA_CMD_CACHE_FLUSH_EXT   0xEA
+#define ATA_CMD_PACKET            0xA0
+#define ATA_CMD_IDENTIFY_PACKET   0xA1
+#define ATA_CMD_IDENTIFY          0xEC
+
+// IDE Command Registers
+#define ATA_REG_DATA       0x00
+#define ATA_REG_ERROR      0x01
+#define ATA_REG_FEATURES   0x01
+#define ATA_REG_SECCOUNT   0x02
+#define ATA_REG_LBA_LO     0x03
+#define ATA_REG_LBA_MID    0x04
+#define ATA_REG_LBA_HI     0x05
+#define ATA_REG_HDDEVSEL   0x06
+#define ATA_REG_COMMAND    0x07
+#define ATA_REG_STATUS     0x07
+
+#define SET_CHS 0xA0
+#define SET_LBA 0xE0
+
+// IDE CONTROL BITS
+#define ATA_CTRL_NIEN      0x02
+#define ATA_CTRL_SRST      0x04
+#define ATA_CTRL_HOB       0x80
+
+// IDE Control Registers
+#define ATA_REG_CONTROL    0x02
+#define ATA_REG_ALTSTATUS  0x02
+
+#define PRIMARY_CMD_BASE   0x01F0
+#define PRIMARY_CTRL_BASE  0x03F4
+
+#define DELAY_CYCLES       10000000
+#define SATA_DELAY(delay) do { for(int x = 0; x < (delay); x++); } while(0);
```

Comment:
looks like a hack. Make this function can then add something in the loop that ensures this does not get optimized out.

Reply:
Will something like this work?

E.g.
```
int i = 0;
while (i < delay) {
    i++;
}
```

Reply:
That will not really change a thing, as the optimized will still concludes this does nothing. A common thing is to loop over NOP:
```
void busy_loop(unsigned int n) {
    while (n-- > 0) {
        __asm__ volatile("nop");
    }
}
```
But that assumes there is a "nop" instruction. There is also a way to really loop over nothing by adding an empty  assembly block that just clains it need the loop variables. So far, no optimized dares to remove this.
```
void busy_loop(unsigned int n) {
    while (n-- > 0) {
        __asm__ volatile("" : "+g" (i) : :);
    }
}
```
But it still leaves the open question, how long this will actually waits. Instead of `SATA_DELAY()`, better use something `usleep()` that at least give some indication how long this is supposed to wait.

Reply:
Out of interest, do you know if there is an attribute that tells the compiler not to optimize?

Reply:
I'm assuming that it just waits for a very small moment or something. The intended counter is 10000000, which on a 3.5GHz core, is about 2ms. Given that it only happens in the init phase, I don't think this warrants a proper `usleep()` implementation.

Reply:
* @nomadeel : We could still call this function `busy_loop_us_sleep()`. The the code using this then would call `busy_loop_us_sleep(2000)` to state that it wants to wait 2ms. The implementation of the function can then loop `5000*n` and say, that this is a crude hack where test have shown that passing 5.000.000 loops on 3.5GHz core are about 1 ms.
* @oliver-wm: see https://gcc.gnu.org/onlinedocs/gcc-10.3.0/gcc/Function-Specific-Option-Pragmas.html#Function-Specific-Option-Pragmas and https://gcc.gnu.org/onlinedocs/gcc-10.3.0/gcc/Common-Function-Attributes.html#Common-Function-Attributes#index-optimize-function-attribute, there is __attribute__((optimize("O0")))` or 
```
#pragma GCC push_options
#pragma GCC optimize ("O0")
... code ...
#pragma GCC pop_options
```